include/revng-c/mlir/Dialect/Clift/IR/CliftStorage.h

```diff
@@ -12,166 +14,109 @@
 
 namespace mlir::clift {
 
-struct StructTypeStorage : public mlir::AttributeStorage {
-public:
-  // Some internal mechanism of mlir have access to the key of the storage
-  // instead of the entire storage or the key using that storage. For that
-  // reason it is best to throw everything related to the type inside the key,
-  // and then override operators to pretend they don't know about the non key
-  // fields. That way when it is needed one can access everything.
-  struct Key {
+template<typename StorageT, typename ValueT>
+class ClassTypeStorage : public mlir::AttributeStorage {
+  struct KeyBase {
     uint64_t ID;
-    llvm::StringRef name;
-    uint64_t Size;
-    Optional<llvm::SmallVector<FieldAttr, 2>> fields;
+    llvm::StringRef Name;
+    std::optional<llvm::SmallVector<FieldAttr, 2>> Fields;
 
     // struct storages are never exposed to the user, they are only used
     // internally to figure out how to create unique objects. only operator== is
     // every used, everything else is handled by hasValue
-    bool operator==(const Key &Other) const { return Other.ID == ID; }
-
-    llvm::hash_code hashValue() const { return llvm::hash_value(ID); }
-
-    [[nodiscard]] bool isInitialized() const { return fields.has_value(); }
-  };
-
-  using KeyTy = Key;
-
-  static llvm::hash_code hashKey(const KeyTy &key) { return key.hashValue(); }
-
-  /// Construct the storage from the type name. Explicitly initialize the
-  /// containedType to nullptr, which is used as marker for the mutable
-  /// component being not yet initialized.
-  StructTypeStorage(uint64_t ID) { TheKey.ID = ID; }
-
-  bool operator==(const Key &Other) const { return this->TheKey == Other; }
-
-  /// Define a construction method for creating a new instance of the storage.
-  static StructTypeStorage *
-  construct(mlir::StorageUniquer::StorageAllocator &allocator,
-            const KeyTy &Key) {
-    auto ToReturn = new (allocator.allocate<StructTypeStorage>())
-      StructTypeStorage(Key.ID);
-    if (Key.isInitialized()) {
-      auto Res = ToReturn->mutate(allocator, Key.name, Key.Size, *Key.fields);
-      revng_assert(Res.succeeded());
+    friend bool operator==(const KeyBase &LHS, const KeyBase &RHS) {
```

Comment:
Why `friend` now?

Reply:
Ah, because of CRTP I guess?

Reply:
There's not a big difference between member operators and hidden friends.

```CPP
friend bool operator==(T const& lhs, T const& rhs)
{
  return lhs.something == rhs.something;
}
// vs
bool operator==(T const& other)
{
  return this->something == other.something;
}```

Hidden friends are more symmetrical and behave better in some cases where there are implicit conversions in play. Generally hidden friends are recommended over member operators.